File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/argument/ChannelInviteArgument.java
package dev.slne.surf.social.chat.command.argument;

import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.CustomArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.slne.surf.social.chat.object.Channel;
import dev.slne.surf.social.chat.provider.ChannelProvider;

public class ChannelInviteArgument extends CustomArgument<Channel, String> {
  public ChannelInviteArgument(String nodeName) {
    super(new StringArgument(nodeName), info -> {
      Channel channel = Channel.getChannel(info.input());

      if (channel == null || !channel.hasInvite(info.sender())) {
        throw CustomArgumentException.fromMessageBuilder(new MessageBuilder("Unknown channel: ").appendArgInput());
      } else {
        return channel;
      }
    });

    this.replaceSuggestions(ArgumentSuggestions.strings(info -> ChannelProvider.getInstance().getChannels().values().stream().filter(channel -> channel.hasInvite(info.sender())).map(Channel::getName).toArray(String[]::new)));
  }
}

File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/argument/ChannelInviteArgument.kt
package dev.slne.surf.social.chat.command.argument

import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.CustomArgument
import dev.jorel.commandapi.arguments.StringArgument
import dev.slne.surf.social.chat.`object`.Channel
import dev.slne.surf.social.chat.provider.ChannelProvider

class ChannelInviteArgument(nodeName: String) :
    CustomArgument<Channel, String>(
        StringArgument(nodeName),
        { info ->
            val channel = Channel.getChannel(info.input())
                ?: throw CustomArgumentException.fromMessageBuilder(MessageBuilder("Unknown channel: ").appendArgInput())

            if (!channel.hasInvite(info.sender())) {
                throw CustomArgumentException.fromMessageBuilder(MessageBuilder("Unknown channel: ").appendArgInput())
            }
            channel
        }) {
    init {
        this.replaceSuggestions(ArgumentSuggestions.stringCollection { info ->
            ChannelProvider.channels.values
                .filter { it.hasInvite(info.sender()) }
                .map { it.name }
            }
        )
    }
}